sklearn/cluster/hierarchical.py

```diff
@@ -704,9 +734,17 @@ def fit(self, X, y=None):
                              " instance, got 'memory={!r}' instead.".format(
                                  type(memory)))
 
-        if self.n_clusters <= 0:
+        self.n_clusters_ = self.n_clusters
+
+        if self.n_clusters_ is not None and self.n_clusters_ <= 0:
```

Comment:
why aren't you just using `self.n_clusters` here? `self.n_clusters` should be the user's input, and should be used in validation. `n_clusters_` should be output to the user.